### PR TITLE
feat: dashboard filters with image format

### PR DESCRIPTION
**Pull request by joaoviana**
Relates to: #4425

### Description:

Adds the ability to save `filters` into a scheduler
Overrides dashboard filters on Minimal page aka scheduler page for it to take a screenshot of. 

- [x] There is a third tab in the scheduled delivery modal called `filters`
- [x] The filters tab includes a list of all of the existing dashboard filters
- [x] You cannot add more filters or delete filters from the dashboard, only adjust the values. 
- [x] Adjusting the filter has the same behaviour as adjusting saved filters in a dashboard. E.g.:
  - [x] if the filter is applied to different fields on different tiles, then this is still the case for the scheduled delivery
  - [x] all the filter conditions are joined using `AND` 
  - [x] You can override the operator
  - [x] you can override the value
  - [x] you CANNOT override the field that's being filtered
- [x] temporary filters that you add in `view` mode are _**not**_ included _**or**_ able to be adjusted in the scheduled delivery  (they _**have**_ to be saved filters)
- [x] If you remove a saved filter from a dashboard, then it would also get removed fromt eh scheduled delivery (even if you overrode it or something)

**note**: This doesn't work with `Send now` feature yet

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx [hunk 1]
```diff
@@ -29,7 +29,7 @@ import {
 import { useProject } from '../../../hooks/useProject';
 import { useDashboardContext } from '../../../providers/DashboardProvider';
 
-type SchedulerFilterRule = DashboardFilterRule & {
+type SchedulerFilterRule = Omit<DashboardFilterRule, 'tileTargets'> & {
```
Comment (IrakliJani) [inline packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:32]: 👍

**Comment by joaoviana** [comment]
I'm aware of some `z-index` issues with select popovers. Looking into it!

File: packages/frontend/src/pages/MinimalDashboard.tsx [hunk 2]
```diff
@@ -19,12 +24,18 @@ const ResponsiveGridLayout = WidthProvider(Responsive);
...
         data: dashboard,
         isError,
         error,
     } = useDashboardQuery(dashboardUuid);
 
+    const { data: scheduler } = useScheduler(schedulerUuid!, {
+        enabled: !!schedulerUuid,
+    });
+
     if (isError) {
```
Comment (ZeRego) [inline packages/frontend/src/pages/MinimalDashboard.tsx:39]: Non blocking issue: We need to update this condition based on error from fetching the scheduler. We could do it in a separate PR.

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx [hunk 3]
```diff
@@ -166,20 +201,27 @@ const SchedulerFilters: FC<SchedulerFiltersProps> = ({
...
+                    {dashboard?.filters?.dimensions.map((filter) => {
+                        const schedulerFilter =
+                            schedulerFiltersData && schedulerFiltersData.length
+                                ? schedulerFiltersData.find(
+                                      (f) =>
+                                          f.target.fieldId ===
+                                              filter.target.fieldId &&
+                                          f.target.tableName ===
+                                              filter.target.tableName,
+                                  )
```
Comment (ZeRego) [inline packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:214]: In `applyDimensionOverrides` and in line 130 in this file, we use the filter id to find a match. Here we use the field ID  and table name. 

This will be a problem if:

- create dashboard with filter A ( with ID X + field ID 1 and table name W)
- create scheduler with filter B (with ID X + field ID 1 and table name W )  --> scheduler filter is valid
- update dashboard with no filters  --> scheduler filter is invalid
- update dashboard with filter C (with ID Y + field ID 1 and table name W) --> scheduler filter is shown since field id and table name match, but onChange returns undefined since the filter ids don't match

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx [hunk 4]
```diff
@@ -100,51 +100,86 @@ const FilterItem: FC<SchedulerFilterItemProps> = ({
...
+            if (hasChanged) {
+                if (isExistingFilter) {
+                    updatedFilters[filterIndex] = schedulerFilter;
+                } else {
+                    updatedFilters.push(schedulerFilter);
+                }
+            }
+            if (isReverted) {
+                updatedFilters.splice(filterIndex, 1);
+            }
```
Comment (ZeRego) [inline packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:171]: Note: found it hard to review this logic because we are creating variables and making calculations ahead of time. I had to go back and forward in the code to recreate all the different scenarios. I still don't know if there is an edge case where both `hasChanged` and `isReverted` can be `true`


eg: if we know they reverted the filter value, we don't need the `filterToCompareAgainst`, `hasChanged` etc

I wonder if we could have a `if else` where all the logic for `hasChanged` would be inside the else block

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx [hunk 5]
```diff
@@ -100,51 +101,86 @@ const FilterItem: FC<SchedulerFilterItemProps> = ({
...
-            setSchedulerFilters(newState);
+            if (!originalFilter) return;
 
-            // TODO: sync with upper component
-            // call onChange - this can be debounced
-        },
-        [schedulerFilters],
-    );
+            let updatedFilters = schedulerFiltersData
+                ? [...schedulerFiltersData]
```
Comment (ZeRego) [inline packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:137]: Is there a place where we remove the invalid scheduler filters and then save? 
Looking at the code it looks like we don't show the invalid filters but they will keep poluting the database.

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx (hunk 3 above)
Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:214]: Ah, I see. Makes total sense to not use this `fieldId` and `tableName` - and to avoid dashboard filters updates + changes to fields with diff operators, values, that could cause unexpected results. Changing it to search by `id` so we ensure uniqueness!

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx (hunk 5 above)
Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:137]: great point, going to replicate that and find a solution. I think this is not handled yet!

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx [hunk 6]
```diff
@@ -79,6 +79,7 @@ const FilterItem: FC<SchedulerFilterItemProps> = ({
                             tileTargets: undefined,
                         });
                     }}
+                    withinPortal
```
Comment (ZeRego) [inline packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:82]: We need the same fix for `<FilterInputComponent`

Screenshot 2023-11-20 at 15 44 14

Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:82]: exactly! thanks 😅

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx (hunk 4 above)
Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx]: Yea, agree. This is def not that straightforward. I've separated this logic into 2 functions: one that handles the `onChange` and a middle one for updating the fields and getting the correct types + updating/removing the changes. 
Also, the check if `isReverted` has been moved further up and it returns so that we avoid extra calculations when that's the case. 

I'm also not sure if there is such a case: reverted **and** changed. I think it's not possible atm

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx (hunk 6 above)
Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx:79]: done :https://github.com/lightdash/lightdash/pull/7985/commits/e6314449726d23b5e5775c6679d88fb1c5bc1a3d

File: packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx (hunk 5 above)
Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx]: done: https://github.com/lightdash/lightdash/pull/7985/commits/7088c21f0c0364d08a963fdcadef9fd34b98b866

there's also the possibility that the user doesn't interact with the UI, but the scheduler keeps going. _old_/invalid filters should then be removed on the BE through some type of logic. But I think that can be handled on a different PR

Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx]: I've also reordered some conditional checks to make it more readable but still allow `schedulerFilters` to be `undefined` if no change etc etc [chore: reorder conditional checks on scheduler Filters](https://github.com/lightdash/lightdash/pull/7985/commits/58b4811894e910a1d5e194888185ecaae5524132)

**Comment by joaoviana** [comment]
Tested: 
* dashboard filter -> override on scheduler -> done
* dashboard filter -> remove filter that was being overridden -> done

**Review by ZeRego — APPROVED** [review]
